lib/ai/model/tools/auditquery.go

```diff
@@ -0,0 +1,151 @@
+package tools
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sashabaranov/go-openai"
+
+	"github.com/gravitational/teleport/gen/go/eventschema"
+	"github.com/gravitational/teleport/lib/ai/tokens"
+)
+
+type AuditQueryGenerationTool struct {
+	LLM *openai.Client
+}
+
+func (t *AuditQueryGenerationTool) Name() string {
+	return "Audit Query Generation"
+}
+
+func (t *AuditQueryGenerationTool) Description() string {
+	return `Generates a SQL query that can be ran against teleport audit events.
+The input must be a single string describing what the query must achieve.`
+}
+
+func (t *AuditQueryGenerationTool) Run(_ context.Context, _ *ToolContext, _ string) (string, error) {
+	// This is stubbed because AuditQueryGenerationTool is handled specially.
+	// This is because execution of this tool breaks the loop and returns a command suggestion to the user.
+	// It is still handled as a tool because testing has shown that the LLM behaves better when it is treated as a tool.
+	//
+	// In addition, treating it as a Tool interface item simplifies the display and prompt assembly logic significantly.
+	return "", trace.NotImplemented("not implemented")
+}
+
+func (t *AuditQueryGenerationTool) ChooseEventTable(ctx context.Context, input string, tc *tokens.TokenCount) (string, error) {
+	tableList, err := eventschema.QueryableEventList()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	prompt := []openai.ChatCompletionMessage{
+		{
+			Role: openai.ChatMessageRoleSystem,
+			Content: `You are a tool that find the correct table to run a query on.
+You will be given a list of tables, and a request from the user.
+You MUST RESPOND ONLY with a single table name. If no table can answer the question, respond 'none'.`,
+		},
+		{
+			Role:    openai.ChatMessageRoleUser,
+			Content: tableList,
+		},
+		{
+			Role:    openai.ChatMessageRoleUser,
+			Content: fmt.Sprintf("The user request is: %s", input),
+		},
+	}
+	promptTokens, err := tokens.NewPromptTokenCounter(prompt)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	tc.AddPromptCounter(promptTokens)
+
+	response, err := t.LLM.CreateChatCompletion(
+		ctx,
+		openai.ChatCompletionRequest{
+			Model:       openai.GPT4,
+			Messages:    prompt,
+			Temperature: 0,
+		},
+	)
+
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	completion := response.Choices[0].Message.Content
+	completionTokens, err := tokens.NewSynchronousTokenCounter(completion)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+	tc.AddCompletionCounter(completionTokens)
+
+	eventType := strings.ToLower(completion)
+	if !eventschema.IsValidEventType(eventType) {
+		return "", trace.CompareFailed("Model response is not a valid event type: '%s'", eventType)
+	}
```

Comment:
Here we can implement a retry logic, but it feels like re-implementing the agent logic. WDYT?

Reply:
I think you have 3 options:
1. Leave as it is.
2. Extract the retry logic from the agent and use it here.
3. Refactor the agent to be able to use a custom initial prompt, set lower temperature etc (IMO this will add the most boilerplate).

I don't have a strong opinion here. I prefer less code 🤷‍♂️